src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt

```diff
@@ -0,0 +1,89 @@
+package pt.up.fe.ni.website.backend.service
+
+import org.springframework.http.HttpStatus
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.crypto.password.PasswordEncoder
+import org.springframework.security.oauth2.jwt.JwtClaimsSet
+import org.springframework.security.oauth2.jwt.JwtDecoder
+import org.springframework.security.oauth2.jwt.JwtEncoder
+import org.springframework.security.oauth2.jwt.JwtEncoderParameters
+import org.springframework.stereotype.Service
+import org.springframework.web.server.ResponseStatusException
+import pt.up.fe.ni.website.backend.config.auth.AuthConfigProperties
+import pt.up.fe.ni.website.backend.model.Account
+import java.time.Duration
+import java.time.Instant
+import java.util.stream.Collectors
+
+@Service
+class AuthService(
+    val accountService: AccountService,
+    val authConfigProperties: AuthConfigProperties,
+    val jwtEncoder: JwtEncoder,
+    val jwtDecoder: JwtDecoder,
+    private val passwordEncoder: PasswordEncoder
+) {
+    fun authenticate(email: String, password: String): Account {
+        val account = accountService.getAccountByEmail(email)
+        if (!passwordEncoder.matches(password, account.password)) {
+            throw ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "invalid credentials")
+        }
+        val authentication = UsernamePasswordAuthenticationToken(email, password, getAuthorities(account))
+        SecurityContextHolder.getContext().authentication = authentication
+        return account
+    }
+
+    fun generateAccessToken(account: Account): String {
+        return generateToken(account, Duration.ofMinutes(authConfigProperties.jwtAccessExpirationMinutes))
+    }
+
+    fun generateRefreshToken(account: Account): String {
+        return generateToken(account, Duration.ofDays(authConfigProperties.jwtRefreshExpirationDays), true)
+    }
+
+    fun refreshAccessToken(refreshToken: String): String {
+        val jwt =
+            try {
+                jwtDecoder.decode(refreshToken)
+            } catch (e: Exception) {
+                throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "invalid refresh token")
+            }
```

Comment:
Continuing the comment above, global exception handling also simplifies services in cases like this
```suggestion
        val jwt =  jwtDecoder.decode(refreshToken)
```